torch/sparse/__init__.py

```diff
@@ -201,3 +201,269 @@ def log_softmax(input: Tensor, dim: int, dtype: Optional[DType] = None) -> Tenso
+            result = torch.sparse_coo_tensor(indices, values, shape, device=mask_input.device)
+        else:
+            if mask_input.ndim == 0:
+                # Workaround https://github.com/pytorch/pytorch/issues/65400
+                dim_ = ()
+            else:
+                dim_ = dim
+            result = torch.sparse.sum(mask_input, dim=dim_, dtype=dtype)
+            if result.dtype != dtype:
+                # Workaround https://github.com/pytorch/pytorch/issues/65392
+                result = result.to(dtype=dtype)
+            if result.layout == torch.strided:
+                result = result.to_sparse()
+
+        if keepdim and mask_input.ndim > 0:
+            # torch.sparse.sum does not support keepdim argument, so,
+            # here we restore the squeezed dimensions
+            if mask_input.dense_dim() > 0:
+                raise NotImplementedError('torch.sparse.masked_sum on hybrid COO sparse tensor')
+            dim_ = _canonical_dim(dim, mask_input.ndim)
+            indices = torch.zeros((mask_input.ndim, result._nnz()), dtype=torch.int64, device=mask_input.device)
+            i = 0
+            for d in range(mask_input.ndim):
+                if d in dim_:
+                    continue
+                indices[d] = result._indices()[i]
+                i += 1
+            shape = tuple((1 if i in dim_ else mask_input.shape[i]) for i in range(mask_input.ndim))
+            result = torch.sparse_coo_tensor(indices, result._values(), shape, dtype=result.dtype, device=result.device)
+        return result
+
+    # TODO: elif input.layout == torch.sparse_csr:
+    else:
+        raise NotImplementedError(f'masked_sum of {input.layout} tensor')
+
+
+@_apply_docstring_templates
+def masked_prod(input: Tensor,
+                dim: DimOrDims = None,
+                *,
+                keepdim: bool = False,
+                dtype: Optional[DType] = None,
+                mask: Optional[Tensor] = None) -> Tensor:
+    """
+{masked_reduction_signature}
+
+{masked_reduction_descr}
+
+{masked_reduction_args}
+
+{masked_reduction_example}
+    >>> torch.sparse.masked_prod(input, 1, mask=mask)
+    tensor([3,  1])
+    """
+    if dtype is None:
+        dtype = input.dtype
+    if input.layout == torch.strided:
+        mask_input = input if mask is None else torch.where(mask, input, torch.ones_like(input))
+        if dim is None or dim == ():
+            return torch.prod(mask_input, dtype=dtype)
+        elif isinstance(dim, int):
+            return torch.prod(mask_input, dim, keepdim, dtype=dtype)
+        else:
+            raise NotImplementedError(f'masked_prod of {input.layout} tensor (shape={input.shape}) for non-integer dim={dim}')
+    # TODO: elif input.layout == torch.sparse_coo:
+    # TODO: elif input.layout == torch.sparse_csr:
+    else:
+        raise NotImplementedError(f'masked_prod of {input.layout} tensor')
+
+
+def masked_mask(input: Tensor,
```

Comment:
Can't we use any for this?

Reply:
`torch.any` is restricted to integer `dim`.
Until `torch.any` is generalized, we'll need `masked_mask` for testing masked reduction operations, specifically, for non-integral `dim` and broadcasting mask support.

Reply:
Could I ask you to add a leading underscore then or maybe move it into the tests only? We can then add an issue that requests multiple dimension support for torch.any. Seems like a reasonable feature request and we don't have to add a function that does the same thing, but with slightly more general functionality. Could you submit a feature request to add support for multiple dim for torch.any if you agree?

Reply:
An issue for supporting multiple dim in `torch.any` already exists:  https://github.com/pytorch/pytorch/issues/56586
I am ok with renaming `masked_mask` to `_masked_mask`.
Since output mask is needed for masked-equality-check in several test files and it also handles broadcasting to input shape (that `torch.any` would not do), I would leave `_masked_mask` in `torch.sparse` unless a better place is suggested.

Reply:
Sounds good. Given that we have an official tracking issue and multidim support for our masked reductions is needed for feature parity, it might make sense for us to send a diff that addresses the multidim support for torch.any Given that there's a tracking issue it should be readily accepted. https://github.com/pytorch/pytorch/issues/61582